pymchelper/averaging.py

```diff
@@ -0,0 +1,214 @@
+    Technometrics. 4 (3): 419–420.
+    [2] Schubert, Erich, and Michael Gertz. "Numerically stable parallel computation of (co-) variance."
+    Proceedings of the 30th international conference on scientific and statistical database management. 2018.
+    [3] https://justinwillmert.com/posts/2022/notes-on-calculating-online-statistics/
+    [4] West, D. H. D. (1979). "Updating Mean and Variance Estimates: An Improved Method".
+    Communications of the ACM. 22 (9): 532-535.
+    [5] https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Weighted_incremental_algorithm
+    """
+
+    data: Union[float, ArrayLike] = float('nan')
+
+    _accumulator_S: Union[float, ArrayLike] = field(default=float('nan'), repr=False, init=False)
+    _total_weight_squared: float = field(default=0., repr=False, init=False)
+    total_weight: float = 0
+
+    def update(self, value: Union[float, ArrayLike], weight: float = 1.0, **kwargs):
+        """
+        Update the state of the aggregator with new data.
+        Note that the weights are so called "reliability weights", not frequency weights.
+        If unsure put here the number of histories from the file if you are aggregating data from multiple files.
+        """
+        if weight < 0:
+            raise ValueError("Weight must be non-negative")
+
+        # first pass initialization
+        if not self.updated:
+            self.data = value * 0
+            self._accumulator_S = value * 0
+
+        # W_n = W_{n-1} + w_n
+        self.total_weight += weight
+        self._total_weight_squared += weight**2
+
+        mean_old = self.data
+        # mu_n = (1 - w_n / W_n) * mu_{n-1} + (w_n / W_n) * x_n
+        # or in other words:
+        # mu_n - mu_{n-1} = (w_n / W_n) * (x_n - mu_{n-1})
+        self.data += (weight / self.total_weight) * (value - mean_old)
+
+        self._accumulator_S += weight * (value - self.data) * (value - mean_old)
+
+        self._updated = True
+        logging.debug("Updated aggregator with value %s and weight %s", value, weight)
+
+    @property
+    def mean(self):
+        """Weighted mean of the sample"""
+        return self.data
+
+    @property
+    def variance_population(self):
+        """Biased estimate of the variance"""
+        return self._accumulator_S / self.total_weight
+
+    @property
+    def variance_sample(self):
+        """
+        Unbiased estimate of the variance.
+        The bias of the weighted estimator if (1 - sum w_i^2 / W_n^2), or in other words:
+        1 - "sum of squares of weights" / "square of sum of weights".
+        For all equal weights the bias is 1 - n * w^2/((n * w)^2) = 1 - 1/n which
+        leads to the well known formula for the sample variance.
+        Here we use the weighted version of the formula.
+        """
+        return self._accumulator_S / (self.total_weight - (self._total_weight_squared / self.total_weight))
```

Comment:
check if self.total_weight != 0 ?

Reply:
That could happen only if you haven't updated the aggregator. Good idea to throw a `ValueError` in that case